integration/fuse/src/main/java/alluxio/cli/FuseShell.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.cli;
+
+import alluxio.AlluxioURI;
+import alluxio.Constants;
+import alluxio.client.file.FileSystem;
+import alluxio.client.file.URIStatus;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.exception.status.InvalidArgumentException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Map;
+
+/**
+ * Fuse shell to run Fuse special commands.
+ */
+public final class FuseShell {
+  private static final Logger LOG = LoggerFactory.getLogger(FuseShell.class);
+
+  private final AlluxioConfiguration mConf;
+  private final FileSystem mFileSystem;
+  private final Map<String, Command> mCommands;
+  private AlluxioURI mPath;
+  private String[] mCmds;
+
+  /**
+   * Creates a new instance of {@link FuseShell}.
+   *
+   * @param fs Alluxio file system
+   * @param conf Alluxio configuration
+   */
+  public FuseShell(FileSystem fs, AlluxioConfiguration conf) {
+    mFileSystem = fs;
+    mConf = conf;
+    mCommands = CommandUtils.loadCommands(FuseShell.class.getPackage().getName(),
+        new Class [] {FileSystem.class, AlluxioConfiguration.class},
+        new Object[] {mFileSystem, mConf});
+  }
+
+  /**
+   * Checks if the given uri contains Fuse special command and
+   * gets the path and command info from the uri.
+   *
+   * @param uri check whether the uri contains Fuse special command
+   * @return true if the path is a special path, otherwise false
+   */
+  public boolean validateAndParseURI(AlluxioURI uri) {
+    int index = uri.getPath().lastIndexOf(AlluxioURI.SEPARATOR);
+    if (index == -1) {
+      return false;
+    }
+    String cmdInfo = uri.getPath().substring(index);
+    if (!cmdInfo.startsWith(Constants.ALLUXIO_CLI_PATH)) {
+      return false;
+    }
+    mPath = uri.getParent();
```

Comment:
Fuse shell is not threadsafe, two threads can call getattr and enter here together

Reply:
Got it, thanks for pointing this problem, can I go back to origin version temporary?